Allow instances for users to request configurable consent

machisuji opened the pull request:
WP [#27787](https://community.openproject.com/projects/openproject/work_packages/27787/activity?query_id=1249)

![image](https://user-images.githubusercontent.com/158871/40490664-0e7bf46a-5f64-11e8-86f5-45f313f7343e.png)

**TODOS**

- [x] settings view for changing the consent date
- [x] tests
- [x] Show consent date in user/show

machisuji committed:
user consent WIP

oliverguenther commented on app/controllers/concerns/user_consent.rb:
```diff
@@ -0,0 +1,95 @@
+      consent_finished
+    else
+      render 'account/consent', locals: {
+        consent_info: consent_info[I18n.locale],
```
You might want to fallback here: `consent_info.fetch(I18n.locale, consent_info[:en])` and probably force users to enter at least `:en` when they check consent (or always read a default value)

oliverguenther commented on app/controllers/concerns/user_consent.rb:
```diff
@@ -0,0 +1,95 @@
+  end
+
+  def update_user_consent!(user)
+    user.update consented_at: DateTime.now
```
I think it's safe to bypass all validations here and just use `user.update_column(:consented_at, DateTime.now)` to ensure the user can log in

oliverguenther committed:
Ensure authentication stage is registered during reload

And just for good measure, ensure stages cannot register the same name
twice (does not happen while reloading, but still...)

oliverguenther committed:
Actually check consent checkbox

oliverguenther committed:
Translation for consent

oliverguenther committed:
Replace consent_label with translation

And fall back to existing english version from default.

oliverguenther committed:
Only ask for consent if at least one translation is provided

oliverguenther committed:
Write consent date directly to user

oliverguenther committed:
Rewrite consent

oliverguenther committed:
Add spec

oliverguenther committed:
Add consent information to user/edit

oliverguenther committed:
Better checkbox label

oliverguenther committed:
Add warning when user declines consent

oliverguenther committed:
Fix back_url in init_authentication_stages

machisuji committed:
added consent translations

machisuji committed:
added datetime setting format

machisuji committed:
added generic setting helper

machisuji committed:
added consent date

machisuji committed:
merged remote branch

machisuji committed:
fixed typo

oliverguenther committed:
Run stage on activation as well

oliverguenther committed:
Correct date writing for consent_date

oliverguenther committed:
Add check box to update consent time to current time

oliverguenther committed:
Show consent in account/register form

oliverguenther committed:
Fix other localizations

[ci skip]
